bin/varnishd/cache/cache_hash.c

```diff
@@ -918,27 +918,36 @@ HSH_Unbusy(struct worker *wrk, struct objcore *oc)
 	CHECK_OBJ_NOTNULL(wrk, WORKER_MAGIC);
 	CHECK_OBJ_NOTNULL(oc, OBJCORE_MAGIC);
 	CHECK_OBJ_NOTNULL(oc->boc, BOC_MAGIC);
+	AN(oc->flags & OC_F_BUSY);
 
 	oh = oc->objhead;
 	CHECK_OBJ(oh, OBJHEAD_MAGIC);
+
+	/* NB: It is guaranteed that exactly one request is waiting for
+	 * the objcore for pass objects. The other reference is held by
+	 * the current fetch task.
+	 */
+	if (oh == private_oh) {
+		assert(oc->refcnt == 2);
+		oc->flags &= ~OC_F_BUSY;
```

Comment:
The clearing of OC_F_BUSY without holding a OH lock could be a concern. Are there code paths making branches based on OC_F_BUSY that would race? I guess it is true before this patch set and would be a concern. I notice that this patch set will move places like my concern over to solely synchronize on the BOS state. Though I didn't have the time to do a thorough search for places where this could be a concern.

Reply:
The reason why this is safe is precisely because there exactly two references on the objcore, and the second one is a `req` task waiting for the objcore to drop the busy flag.

There will be no cache lookup attempted on the `private_oh` so there is nothing else eligible for a race.